prediction_market_agent/agents/blockchain_coding_agent/functions.py

```diff
@@ -25,47 +28,44 @@ def tavily_search(
     return tavily_search_pmat(query=query)
 
 
-def fetch_read_methods_from_blockscout(contract_address: str) -> t.Any:
-    w3 = OmenConditionalTokenContract().get_web3()
-    if not is_contract(w3, Web3.to_checksum_address(contract_address)):
-        raise ValueError(f"{contract_address=} is not a contract on Gnosis Chain.")
-    read_not_proxy = fetch_read_methods(contract_address)
-    read_proxy = fetch_read_methods_proxy(contract_address)
-    # ToDo - Fetch write methods
-    return read_not_proxy + read_proxy
+def get_rpc_endpoint() -> str:
+    # return "https://rpc.gnosischain.com"
+    return "http://localhost:8545"
 
 
-def fetch_read_methods(contract_address: str) -> list[dict[str, t.Any]]:
-    url = f"https://gnosis.blockscout.com/api/v2/smart-contracts/{contract_address}/methods-read?is_custom_abi=false"
-    r = requests.get(url)
-    data: list[dict[str, t.Any]] = r.json()
-    return data
+def checksum_address(address: str) -> ChecksumAddress:
+    return Web3.to_checksum_address(address)
 
 
 def is_contract(web3: Web3, contract_address: ChecksumAddress) -> bool:
     return bool(web3.eth.get_code(contract_address))
 
 
-def fetch_read_methods_proxy(contract_address: str) -> list[dict[str, t.Any]]:
-    url = f"https://gnosis.blockscout.com/api/v2/smart-contracts/{contract_address}/methods-read-proxy?is_custom_abi=false"
-    r = requests.get(url)
-    data: list[dict[str, t.Any]] = r.json()
-    return data
-
-
-def get_rpc_endpoint() -> str:
-    return RPCConfig().gnosis_rpc_url
-
-
-def checksum_address(address: str) -> ChecksumAddress:
-    return Web3.to_checksum_address(address)
+def fetch_source_code_and_abi_from_contract(
```

Comment:
Any reason to not re-use `persistent_inmemory_cache` that we already have?

```suggestion
@persistent_inmemory_cache
def fetch_source_code_and_abi_from_contract(
```

Then we don't need a new caching logic with `CachedSession`